### PR TITLE
[API compatibility] Support Tensor.to

### PR DESCRIPTION
**Pull request by zeroRains**
### PR Category

User Experience

### PR Types

New features

### Description

为paddle.Tensor.to添加non_blocking和copy参数，使其对齐torch.Tensor.to()。

在PaConvert下自测，paddle.Tensor.to与torch.Tensor.to对齐。

pcard-71500

image

### CODE REVIEW

File: python/paddle/base/dygraph/tensor_patch_methods.py [hunk 1]
```diff
@@ -850,7 +862,11 @@ def get_device_dtype_from_tensor(other):
                 device, dtype = get_device_dtype_from_tensor(
                     kwargs.get("other", None)
                 )
-        return self._to(device, dtype, blocking)
+        blocking = False if not blocking or non_blocking else True
+        res = self._to(device, dtype, blocking)
+        if copy_tensor:
```
Comment (zhwesky2010) [inline python/paddle/base/dygraph/tensor_patch_methods.py]: 这个可能导致需要copy两次的问题，对性能影响较大。
在_to里面copy了就不用处理了，在里面没copy才需要处理，这个flag可以传过去，跟里面的分支一起判断。

File: python/paddle/pir/math_op_patch.py [hunk 2]
```diff
@@ -1295,8 +1316,13 @@ def dispatch_to_signature(*args, **kwargs):
             args["dtype"] = other.dtype
             # in dy2static, we need show warning for this case
             other.place  # noqa: B018
-
-        return self._to(**args)
+        args["blocking"] = (
+            False if not args.get("blocking", False) or non_blocking else True
+        )
+        res = self._to(**args)
+        if copy_tensor:
```
Comment (zhwesky2010) [inline python/paddle/pir/math_op_patch.py]: 同上，导致copy两次的问题

**Comment by zhwesky2010** [comment]
Tensor.to的non_blocking支持位置参数用法吗？

支持的话会有像Tensor.copy_一样的问题，即同一个位置既可能是 blocking，也可能是non_blocking。

**Comment by zeroRains** [comment]
> Tensor.to的non_blocking支持位置参数用法吗？
> 
> 支持的话会有像Tensor.copy_一样的问题，即同一个位置既可能是 blocking，也可能是non_blocking。

blocking支持的，不过non_blocking和copy只支持字典参数，blocking和non_blocking参数冲突时，和copy的处理逻辑一样，默认使用blocking，只有blocking为false或者non_blocking为true时，才会使用non_blocking

File: python/paddle/pir/math_op_patch.py [hunk 3]
```diff
@@ -1106,26 +1108,41 @@ def transform(t, device, dtype, blocking):
...
+
+    def __deepcopy__(self, memo):
+        if isinstance(self.place, paddle.CUDAPlace):
+            new_tensor = _C_ops.memcpy(self, 1)
+        elif isinstance(self.place, paddle.CUDAPinnedPlace):
+            new_tensor = _C_ops.memcpy(self, 2)
+        else:
+            new_tensor = _C_ops.memcpy(self, 0)
+        memo[id(self)] = new_tensor
+        return new_tensor
```
Comment (zeroRains) [inline python/paddle/pir/math_op_patch.py]: image

在静态图里，通过copy.deepcopy调用的__deepcopy__方法中，self.place为None，导致只能复制到cpu上，这是符合预期的吗？

File: python/paddle/pir/math_op_patch.py [hunk 4]
```diff
@@ -1106,26 +1108,41 @@ def transform(t, device, dtype, blocking):
...
-
+            if copy_tensor:
+                return copy.deepcopy(new_t)
             return new_t
 
-        return transform(self, device, dtype, blocking)
+        return transform(self, device, dtype, blocking, copy_tensor)
+
+    def __deepcopy__(self, memo):
+        if isinstance(self.place, paddle.CUDAPlace):
```
Comment (zhwesky2010) [inline python/paddle/pir/math_op_patch.py]: 静态图在编译期是没有place的，在运行期才有place。可以试下 `x.clone().detach()`

**Review by zhwesky2010 — APPROVED** [review]

File: python/paddle/pir/math_op_patch.py [hunk 5]
```diff
@@ -1106,26 +1108,36 @@ def transform(t, device, dtype, blocking):
...
-
+            if copy_tensor:
+                return copy.deepcopy(new_t)
             return new_t
 
-        return transform(self, device, dtype, blocking)
+        return transform(self, device, dtype, blocking, copy_tensor)
+
+    def __deepcopy__(self, memo):
+        new_tensor = self.clone().detach()
```
Comment (zhwesky2010) [inline python/paddle/pir/math_op_patch.py:1138]: 刚看了下，这个方法会改变x的stop_gradient 属性。

所以用`x.clone().detach()` 还有点问题，还得看看有没其他更好的 copy 方式。
